akka-actor-typed/src/main/scala/akka/actor/typed/receptionist/Receptionist.scala

```diff
@@ -110,24 +129,121 @@ object Receptionist extends ExtensionId[Receptionist] {
 
+  /**
+   * Internal API
+   */
+  @InternalApi
+  object MessageImpls {
+    // some trixery here to provide a nice _and_ safe API in the face
+    // of type erasure, more type safe factory methods for each message
+    // is the user API below while still hiding the type parameter so that
+    // users don't incorrecly match against it
+
+    final case class Register[T] private[akka] (
+      key:             ServiceKey[T],
+      serviceInstance: ActorRef[T],
+      replyTo:         Option[ActorRef[Receptionist.Registered]]) extends Command
+
+    final case class Registered[T] private[akka] (key: ServiceKey[T], _serviceInstance: ActorRef[T]) extends Receptionist.Registered {
+      def isForKey(key: ServiceKey[_]): Boolean = key == this.key
+      def serviceInstance[M](key: ServiceKey[M]): ActorRef[M] = {
+        if (key != this.key) throw new IllegalArgumentException(s"Wrong key [$key] used, must use listing key [${this.key}]")
+        _serviceInstance.asInstanceOf[ActorRef[M]]
+      }
+
+      def getServiceInstance[M](key: ServiceKey[M]): ActorRef[M] =
+        serviceInstance(key)
+    }
+
+    final case class Find[T] private[akka] (key: ServiceKey[T], replyTo: ActorRef[Receptionist.Listing]) extends Command
+
+    final case class Listing[T] private[akka] (key: ServiceKey[T], _serviceInstances: Set[ActorRef[T]]) extends Receptionist.Listing {
+
+      def isForKey(key: ServiceKey[_]): Boolean = key == this.key
+
+      def serviceInstances[M](key: ServiceKey[M]): Set[ActorRef[M]] = {
+        if (key != this.key) throw new IllegalArgumentException(s"Wrong key [$key] used, must use listing key [${this.key}]")
+        _serviceInstances.asInstanceOf[Set[ActorRef[M]]]
+      }
+
+      def getServiceInstances[M](key: ServiceKey[M]): java.util.Set[ActorRef[M]] =
+        serviceInstances(key).asJava
+    }
+
+    final case class Subscribe[T] private[akka] (key: ServiceKey[T], subscriber: ActorRef[Receptionist.Listing]) extends Command
+
+  }
+
   /**
    * Associate the given [[akka.actor.typed.ActorRef]] with the given [[ServiceKey]]. Multiple
    * registrations can be made for the same key. De-registration is implied by
    * the end of the referenced Actor’s lifecycle.
    *
-   * Registration will be acknowledged with the [[Registered]] message to the given replyTo actor.
+   * Registration will be acknowledged with the [[Registered]] message to the given replyTo actor
+   * if there is one.
    */
-  final case class Register[T](key: ServiceKey[T], serviceInstance: ActorRef[T], replyTo: ActorRef[Registered[T]]) extends Command
   object Register {
-    /** Auxiliary constructor to be used with the ask pattern */
-    def apply[T](key: ServiceKey[T], service: ActorRef[T]): ActorRef[Registered[T]] ⇒ Register[T] =
-      replyTo ⇒ Register(key, service, replyTo)
+
+    /**
+     * Create a Register without Ack that the service was registered
+     */
+    def apply[T](key: ServiceKey[T], service: ActorRef[T]): Command =
```

Comment:
should we define traits for `Register`, `Find`, `Subscribe` instead of using `Command`?

Reply:
I'm not sure that is very useful, it could be more consistent, but those types will never be used anywhere as the receptionist deals with the message implementations.